### ISSUE TITLE
XISO needs to be trimmed at end

### ISSUE THREAD

**Issue by Deterous** [Qubits01/redump2xiso#1]
I tried your tool, however it only splits the XISO at the beginning of the game partition but does not trim the filler and L1 video portion from the end of the XISO. As a result, the XISO is larger than it needs to be, and the video ISO is not complete.

I've attempted to make a script in dotnet that does the above: https://github.com/Deterous/XboxKit
However it does not (yet?) convert the video ISO and XISO back to redump ISO.

**Comment by Qubits01** [comment]
I don't know what you mean exactly. XISO does not contain the video partition. that's exactly what gets split off.

**Comment by Deterous** [comment]
In redump ISOs, the video partition is split and located in two places: at the beginning of the ISO, and at the end of the ISO. This is because the video partition is split across the two layers of the disc. Hopefully this image helps you visualise it:
![image](https://github.com/user-attachments/assets/a14eb9a8-37a6-4a81-85cf-cbf593d55e0b)
(Note: redump ISO does not contain Middle Zone B or C, it is contiguous at the actual layerbreak)

When running your program, it only splits the ISO at the start of L0 game, meaning the video partition contains L0 video and Middle Zone A. It results in the XISO containing Middle Zone D and L1 video.

**Comment by Qubits01** [comment]
I see. Did you investigate this further? For OG Xbox images, starting from offset 0x1bb0b0000 to the end of the file (373MB) the whole area is just Zeros. Could that be it, or does it start earlier. It's been too long that I worked on this. But I think up to offset 0x1bb0b0000 there is padding data present, so that should be it. Just strange that it's just zeros.
I checked the split firt part until L0 video and the video seems to play completely with nothing missing...

**Comment by Deterous** [comment]
Yes, it is padded with zeroes (Middle Zone D). Then it ends with a small number of sectors for the L1 portion of the video partition.
For original xbox discs it probably contains no useful data, but there are definitely non-zero bytes near the end of the file.
My code correctly merges the two sections: https://github.com/Deterous/XboxKit/blob/main/Program.cs
It also just drops the filler / zeroes completely

**Comment by Qubits01** [comment]
Thank you for the information. I will try to adapt my script to the info you gave when I find the time.

**Comment by mokMan23** [comment]
> Thank you for the information. I will try to adapt my script to the info you gave when I find the time.

Are you still planning to update this tool? I've been looking for ways to compress my Redump ISOs more efficiently. I tried both @Deterous' tool & yours, but neither give me the compression ratios I want compared to a "rebuilt" XISO. Is there any way I could get better compression ratios with a reversible method, or should I cut my losses & stick with "traditional" XISOs?

**Comment by Deterous** [comment]
> > Thank you for the information. I will try to adapt my script to the info you gave when I find the time.
> 
> Are you still planning to update this tool? I've been looking for ways to compress my Redump ISOs more efficiently. I tried both [@Deterous](https://github.com/Deterous)' tool & yours, but neither give me the compression ratios I want compared to a "rebuilt" XISO. Is there any way I could get better compression ratios with a reversible method, or should I cut my losses & stick with "traditional" XISOs?

My tool (and probably Qubits01's tool as well) do not compress the XISO at all. Rather, I just trim the ISO such that only the native XISO partition remains. The video partition and the filler middle zones are removed. Properly designed Xbox emulators should be able to read this native format, as it represents exactly what is on the disc. 

[extract-xiso](https://github.com/XboxDev/extract-xiso) and similar programs "rebuild" the XISO as you say, moving the files to a new filesystem (that can compress much more due to the removal of the junk sectors). It is completely different from what is on the disc, but it should remain playable in an emulator (except in rare circumstances, which may require the original filesystem/disc image).

I wrote my script as a proof of concept to help with development for [NKit2](https://github.com/Nanook/NKit/wiki), a toolkit (for more than just Nintendo systems) that among other things aims to do things that you want to do, properly reversible ISO<->XISO with full compression. The junk sectors would have to be stored separately for the reverse process, similar to the video partition. While NKit2 supports many features of the XISO filesystem such as scanning and extracting files, it doesn't yet do what you want.

While it's not reversible, the junk sectors could also be zeroed which means the file would compress much better, while retaining the original filesystem and the same original file size.

**Comment by mokMan23** [comment]
> > > Thank you for the information. I will try to adapt my script to the info you gave when I find the time.
> > 
> > 
> > Are you still planning to update this tool? I've been looking for ways to compress my Redump ISOs more efficiently. I tried both [@Deterous](https://github.com/Deterous)' tool & yours, but neither give me the compression ratios I want compared to a "rebuilt" XISO. Is there any way I could get better compression ratios with a reversible method, or should I cut my losses & stick with "traditional" XISOs?
> 
> My tool (and probably Qubits01's tool as well) do not compress the XISO at all. Rather, I just trim the ISO such that only the native XISO partition remains. The video partition and the filler middle zones are removed. Properly designed Xbox emulators should be able to read this native format, as it represents exactly what is on the disc.
> 
> [extract-xiso](https://github.com/XboxDev/extract-xiso) and similar programs "rebuild" the XISO as you say, moving the files to a new filesystem (that can compress much more due to the removal of the junk sectors). It is completely different from what is on the disc, but it should remain playable in an emulator (except in rare circumstances, which may require the original filesystem/disc image).
> 
> I wrote my script as a proof of concept to help with development for [NKit2](https://github.com/Nanook/NKit/wiki), a toolkit (for more than just Nintendo systems) that among other things aims to do things that you want to do, properly reversible ISO<->XISO with full compression. The junk sectors would have to be stored separately for the reverse process, similar to the video partition. While NKit2 supports many features of the XISO filesystem such as scanning and extracting files, it doesn't yet do what you want.
> 
> While it's not reversible, the junk sectors could also be zeroed which means the file would compress much better, while retaining the original filesystem and the same original file size.

Thanks for the info! I tested your XISO tool with xemu & it worked. I did more research & found out this tool is supposed to be used with [xbox_shrinker](https://github.com/Qubits01/xbox_shrinker), a tool that does what I want. I'll keep an eye on NKit 2 & 360 ISO decryption development, but this'll do for now.

**Comment by Deterous** [comment]
> Thanks for the info! I tested your XISO tool with xemu & it worked. I did more research & found out this tool is supposed to be used with [xbox_shrinker](https://github.com/Qubits01/xbox_shrinker), a tool that does what I want. I'll keep an eye on NKit 2 & 360 ISO decryption development, but this'll do for now.

Yes that's right, unfortunately for later games where the seed is not known, you have to separate the junk sectors for it to be reversible. Wiping them in a non-reversible way and then running my script to trim it to just the XISO would be fine, it's the original filesystem in a compressible format.

**Comment by Deterous** [comment]
@mokMan23 I've completed work to wipe and trim both Xbox and Xbox 360 redump ISOs to compressible XISOs that are completely rebuildable to redump ISOs. You can choose which components to extract, even including XGD3 update files from the video ISO which helps a bit as they're shared between discs.
https://github.com/Deterous/XboxKit

Unfortunately for the filler data it is not possible to get the initial seed, so it does need storing if you want to return to redump ISO losslessly. I've done it slightly differently to Qubit's tool, the filler data includes the zeroed security sectors.
